### PR TITLE
Fixed partitioning in set index for ordered Categorical

### LINKED ISSUE #5671: Rows can be dropped when setting index on a categorical with non-lexical order.
When setting the index on a partitioned dataframe to a categorical whose order is not lexical, rows can be lost. That's a mouthful for sure, and I'm not certain that it would be reasonable to expect otherwise, but I have produced an MRP.

```python
import os
import string, random

import pandas as pd
import numpy as np
import dask
import dask.dataframe as dd

print(dask.__version__)

def random_word(len=10):
    return ''.join(np.random.choice(list(string.ascii_letters), len))

rows = 5000
labels = [random_word(np.random.randint(15, 30)) for i in range(rows)]
data = [i for i in range(len(labels))]

df = pd.DataFrame({'ID': labels, 'data': data})


ddf = dd.from_pandas(df, npartitions=10)
print(ddf.shape[0].compute())

# Drop rows
ddf = ddf.categorize(columns=['ID'])
ddf['ID'] = ddf['ID'].cat.as_ordered()
ddf = ddf.set_index('ID')
print(ddf.shape[0].compute())

# Does not drop rows
ddf = dd.from_pandas(df, npartitions=10)
ddf = ddf.categorize(columns=['ID'])

# Order category lexically
cats = ddf['ID'].cat.categories.values
cats = sorted(cats)
ddf = ddf
ddf['ID'] = ddf['ID'].cat.reorder_categories(cats, ordered=True)

ddf.set_index('ID')
print(ddf.shape[0].compute())
```

In my case, the output generated is:

```
2.8.1
5000
1223
5000
```

This doesn't occur when the dataframe is a single partition. Digging around in the documentation makes me think there might be a scalar lexical comparison instead of a categorical compare somewhere. That or I'm not expecting the right behavior!

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes #5671

### CODE REVIEW

File: dask/dataframe/shuffle.py [hunk 1]
```diff
@@ -101,12 +105,32 @@ def set_index(
...
 
+def maybe_categorical_sorted(values, dtype=None):
+    if dtype is None:
+        return sorted(values)
+    return list(pd.Categorical(values, dtype=dtype).sort_values())
+
+
+def maybe_categorical_compare_minmax(mins, maxes, dtype=None):
+    if dtype is None:
+        return all(mx < mn for mx, mn in zip(maxes[:-1], mins[1:]))
```
Comment (mrocklin) [inline dask/dataframe/shuffle.py]: Do we still need these?

File: dask/dataframe/shuffle.py [hunk 2]
```diff
@@ -163,15 +194,27 @@ def set_partition(
         If using the task-based shuffle, the amount of splitting each
         partition undergoes.  Increase this for fewer copies but more
         scheduler overhead.
+    empty_dataframe_detected : bool, default False
+        Whether all divisions were nan.
```
Comment (mrocklin) [inline dask/dataframe/shuffle.py]: Is this not computable from the divisions that are passed in?  I didn't quite follow the logic above.

**Comment by mrocklin** [comment]
Good to go @TomAugspurger ?

**Comment by TomAugspurger** [comment]
Should be good.

**Pull request merged**